ext/atomic/index.md

```diff
@@ -0,0 +1,510 @@
+> reserved namespace as a prefix.
+
+## <a href="#document-structure" id="document-structure" class="headerlink"></a> Document Structure
+
+A document that supports this extension **MAY** include any of the top-level
+members allowed by the base specification with the exception of `data` and
+`included`, which **MUST NOT** be included.
+
+In addition, such a document **MAY** include either of the following members,
+but not both:
+
+- `atomic:operations` - an array of [operation objects](#operation-objects).
+
+- `atomic:results` - an array of [result objects](#result-objects).
+
+If either `atomic:operations` or `atomic:results` is present, the `errors`
+member **MUST NOT** be included in the same document.
+
+### <a href="#operation-objects" id="operation-objects" class="headerlink"></a> Operation Objects
+
+An operation object **MUST** contain the following member:
+
+- `op`: an operation code, expressed as a string, that indicates the type of
+  operation to perform. Valid values include:
+
+  - `"add"`: creates a new resource or relationship
+  - `"update"`: updates a resource or relationship
+  - `"remove"`: removes a resource or relationship
+
+An operation object **MAY** contain either of the following members, but not
+both, to specify the target of the operation:
+
+- `ref`: an object that **MAY** contain any of the following combinations of
+  members:
+
+  - `type` and `id`: to target an individual resource.
+  - `type`, `id`, and `relationship`: to target an individual resource's relationship.
+
+  Note that `local:id` **MAY** be used instead of `id` when used together with
+  the local identity extension.
+
+- `href`: a string that contains a URI-reference [[RFC3986 Section
+  4.1](https://tools.ietf.org/html/rfc3986#section-4.1)] that identifies the
+  target of the operation.
+
+An operation object **MAY** also contain any of the following members:
+
+- `data`: the operation's "primary data".
+
+- `meta`: a [meta object](#document-meta) that contains non-standard
+  meta-information about the operation.
+
+Different members are required for processing different types of operations, as
+described below.
+
+### <a href="#result-objects" id="result-objects" class="headerlink"></a> Result Objects
+
+An operation result object **MAY** contain any of the following members:
+
+- `data`: the "primary data" resulting from the operation.
+
+- `meta`: a [meta object](#document-meta) that contains non-standard
+  meta-information about the result.
+
+An empty result object (`{}`) is acceptable for operations that are not required
+to return `data`.
+
+## <a href="#processing" id="processing" class="headerlink"></a> Processing
+
+All HTTP requests sent with this extension **MUST** be issued with `POST`.
+
+A server **MUST** perform operations in the order they appear in the
+`atomic:operations` array.
+
+A server **MUST** perform all operations atomically, so that a failure
```

Comment:
Suggested change:
A server **MUST** perform all operations as a single atomic transaction, so that […].

I think we should mention the word "transaction" or "transactional" at least once to make it clear that this is not about processing multiple atomic operations, but about processing multiple operations as one atomic transaction.

Reply:
@ahx thanks for bringing this up. I agree that this line could be clearer. What would you think about the following alternative?

> A server **MUST** perform all operations together atomically, so that [...]

I hope this achieves the effect you're going while avoiding the word "transaction". I actually deliberately avoided using the word "transaction" in this extension, because it brings to mind [ACID](https://en.wikipedia.org/wiki/ACID) guarantees, of which atomicity is just one. And really, the guarantees made by this extension just focus on atomicity. There is no more of a "durability" guarantee than provided by the base spec, for instance. 

On the other hand, I think that using lower-case-t "transaction" in this one instance wouldn't be terribly confusing.